src/backend/address_space_core.h

```diff
@@ -57,45 +51,77 @@ namespace snmalloc
     }
 
     /**
-     * Adds a block to `ranges`.
+     * Set next pointer for a power of two address range.
+     *
+     * This abstracts the use of either
+     * - the pagemap; or
+     * - the first pointer word of the block
+     * to store the next pointer for the list of unused address space of a
+     * particular size.
      */
-    template<SNMALLOC_CONCEPT(ConceptPAL) PAL>
-    void add_block(size_t align_bits, CapPtr<void, CBChunk> base)
+    template<typename Pagemap>
+    void set_next(
+      size_t align_bits,
+      CapPtr<void, CBChunk> base,
+      CapPtr<void, CBChunk> next,
+      Pagemap& pagemap)
     {
-      check_block(base, align_bits);
-      SNMALLOC_ASSERT(align_bits < 64);
-      if (ranges[align_bits][0] == nullptr)
+      if (align_bits >= MIN_CHUNK_BITS)
       {
-        // Prefer first slot if available.
-        ranges[align_bits][0] = base;
+        // The pagemap stores MetaEntrys, abuse the metaslab field to be the
+        // next block in the stack of blocks.
+        MetaEntry t(reinterpret_cast<Metaslab*>(next.unsafe_ptr()), nullptr, 0);
+        pagemap.set(address_cast(base), t);
         return;
       }
 
-      if (ranges[align_bits][1] != nullptr)
+      *(reinterpret_cast<CapPtr<void, CBChunk>*>(base.unsafe_ptr())) = next;
```

Comment:
I fiddled a bit with adding a `struct FreeChunk { CapPtr<FreeChunk, CBChunk> next; };`, which renders this line `base->next = next;`, which is great, and similarly for line 100 below, which just becomes `return base->next;`, but there get to be a lot of casts elsewhere (if you're curious, I've [attached a diff](https://github.com/microsoft/snmalloc/files/6868415/snmalloc-address_space_core-FreeChunk.txt)).  As somewhat sketchy as this line feels, it's probably fine.

Reply:
That looks much nicer